autodiff/forward/real/real.hpp

```diff
@@ -0,0 +1,835 @@
+    For<1, N + 1>([&](auto i) constexpr {
+        expx[i] = Sum<0, i>([&](auto j) constexpr {
+            constexpr auto c = BinomialCoefficient<i.index - 1, j.index>;
+            return c * x[i - j] * expx[j];
+        });
+    });
+    return expx;
+}
+
+template<size_t N, typename T>
+constexpr auto log(const Real<N, T>& x)
+{
+    assert(x[0] != 0 && "autodiff: log(x) has undefined value and derivatives when x = 0");
+    Real<N, T> logx;
+    logx[0] = log(x[0]);
+    For<1, N + 1>([&](auto i) constexpr {
+        logx[i] = x[i] - Sum<1, i>([&](auto j) constexpr {
+            constexpr auto c = BinomialCoefficient<i.index - 1, j.index - 1>;
+            return c * x[i - j] * logx[j];
+        });
+        logx[i] /= x[0];
+    });
+    return logx;
+}
+
+template<size_t N, typename T>
+constexpr auto log10(const Real<N, T>& x)
+{
+    assert(x[0] != 0 && "autodiff: log10(x) has undefined value and derivatives when x = 0");
+    const auto ln10 = 2.302585092994046;
+    Real<N, T> res = log(x);
+    return res /= ln10;
+}
+
+template<size_t N, typename T>
+constexpr auto sqrt(const Real<N, T>& x)
+{
+    Real<N, T> res;
+    res[0] = sqrt(x[0]);
+
+    if constexpr (N > 0)
+    {
+        assert(x[0] != 0 && "autodiff: sqrt(x) has undefined derivatives when x = 0");
+        Real<N, T> a;
+        For<1, N + 1>([&](auto i) constexpr {
+            a[i] = x[i] - Sum<1, i>([&](auto j) constexpr {
+                constexpr auto c = BinomialCoefficient<i.index - 1, j.index - 1>;
+                return c * x[i - j] * a[j];
+            });
+            a[i] /= x[0];
+
+            res[i] = 0.5 * Sum<0, i>([&](auto j) constexpr {
+                constexpr auto c = BinomialCoefficient<i.index - 1, j.index>;
+                return c * a[i - j] * res[j];
+            });
+        });
+    }
+    return res;
+}
+
+template<size_t N, typename T>
+constexpr auto pow(const Real<N, T>& x, const Real<N, T>& y)
+{
+    Real<N, T> res;
+    res[0] = pow(x[0], y[0]);
+    if constexpr (N > 0)
+    {
+        assert(x[0] != 0 && "autodiff: pow(x, y) has undefined derivatives when x = 0");
+        Real<N, T> lnx = log(x);
+        Real<N, T> a;
+        For<1, N + 1>([&](auto i) constexpr {
+            a[i] = Sum<0, i + 1>([&](auto j) constexpr {
+                constexpr auto c = BinomialCoefficient<i.index, j.index>;
+                return c * y[i - j] * lnx[j];
+            });
+
+            res[i] = Sum<0, i>([&](auto j) constexpr {
+                constexpr auto c = BinomialCoefficient<i.index - 1, j.index>;
+                return c * a[i - j] * res[j];
+            });
+        });
+    }
+    return res;
+}
+
+template<size_t N, typename T, typename U, EnableIf<isNumber<U>>...>
+constexpr auto pow(const Real<N, T>& x, const U& c)
+{
+    Real<N, T> res;
+    res[0] = pow(x[0], static_cast<T>(c));
+    if constexpr (N > 0)
+    {
+        assert(x[0] != 0 && "autodiff: pow(x, y) has undefined derivatives when x = 0");
```

Comment:
Actually, pow has derivative at zero (for c >= 1) could you give me a quick explanation? (Sorry, I have no much time to underlying math).

Reply:
See the equation:

![image](https://user-images.githubusercontent.com/5825588/68692217-61e60b00-0575-11ea-9587-7dbc5e68ce02.png)

In the specific case *y* is a constant, and not a `Real<N, T>` number, then we can compute the derivative when `x = 0` and `c >= N`, where `N` is the order of the `Real` number. Otherwise, there will be (at some point) a negative exponent on `0`. 

I think we should consider `pow(x, y) === exp(y * log(x))`, independently if `x` or `y` are constants. Note, however, that we do not compute `pow(x, y)` with `exp(y * log(x))`.